test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -109,310 +113,386 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
+				err := tryFunc(cleanupSteps[i])
+				framework.ExpectNoError(err, "while running cleanup steps")
 			}
-		})
-
-		ginkgo.By("starting a pod to use the claim")
-		command := "echo 'hello world' > /mnt/test/data"
-		l.pod = StartInPodWithVolume(l.cs, l.pvc.Namespace, l.pvc.Name, "pvc-snapshottable-tester", command, config.ClientNodeSelection)
-		l.cleanupSteps = append(l.cleanupSteps, func() {
-			StopPod(l.cs, l.pod)
-		})
 
-		err = e2epv.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, l.cs, l.pvc.Namespace, l.pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
-		framework.ExpectNoError(err)
-
-		ginkgo.By("checking the claim")
-		// Get new copy of the claim
-		l.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.pvc.Namespace).Get(context.TODO(), l.pvc.Name, metav1.GetOptions{})
-		framework.ExpectNoError(err)
-
-		// Get the bound PV
-		_, err = l.cs.CoreV1().PersistentVolumes().Get(context.TODO(), l.pvc.Spec.VolumeName, metav1.GetOptions{})
-		framework.ExpectNoError(err)
-	}
-	cleanup := func(l *snapshottableLocal) {
-		// Depending on how far the test executed, cleanup accordingly
-		// Execute in reverse order, similar to defer stack
-		for i := len(l.cleanupSteps) - 1; i >= 0; i-- {
-			err := tryFunc(l.cleanupSteps[i])
-			framework.ExpectNoError(err, "while running cleanup steps")
 		}
-
-		// All tests will require these driver cleanup tests
-		err := tryFunc(l.driverCleanup)
-		l.driverCleanup = nil
-		framework.ExpectNoError(err, "while cleaning up driver")
-	}
-
-	ginkgo.It("should create snapshot with delete policy [Feature:VolumeSnapshotDataSource]", func() {
-		l := &snapshottableLocal{}
-
-		init(l)
-		defer cleanup(l)
-
-		TestSnapshottable(l, SnapshotClassTest{
-			DeletionPolicy: "Delete",
+		ginkgo.BeforeEach(func() {
+			init()
 		})
-		TestSnapshotDeleted(l, SnapshotClassTest{
-			DeletionPolicy: "Delete",
+		ginkgo.AfterEach(func() {
+			cleanup()
 		})
 
-	})
-
-	ginkgo.It("should not delete snapshot with retain policy [Feature:VolumeSnapshotDataSource]", func() {
-		l := &snapshottableLocal{}
-
-		init(l)
-		defer cleanup(l)
-
-		TestSnapshottable(l, SnapshotClassTest{
-			DeletionPolicy: "Retain",
-		})
-		TestSnapshotDeleted(l, SnapshotClassTest{
-			DeletionPolicy: "Retain",
+		ginkgo.Context("", func() {
+			var (
+				vs        *unstructured.Unstructured
+				vscontent *unstructured.Unstructured
+				vsc       *unstructured.Unstructured
+			)
+
+			ginkgo.BeforeEach(func() {
+				sr := CreateSnapshotResource(sDriver, config, pattern, pvc.GetName(), pvc.GetNamespace())
+				vs = sr.Vs
+				vscontent = sr.Vscontent
+				vsc = sr.Vsclass
+				cleanupSteps = append(cleanupSteps, func() {
+					framework.ExpectNoError(sr.CleanupResource())
+				})
+			})
+
+			ginkgo.It("should delete the VolumeSnapshotContent according to its deletion policy", func() {
+				err = DeleteAndWaitSnapshot(dc, vs.GetNamespace(), vs.GetName(), framework.Poll, framework.SnapshotDeleteTimeout)
```

Comment:
For a retain test case wont this function return err != nil ? line 210 checks unconditionally ExpectNoError?

Reply:
to me this looks simpler (because DeleteAndWaitSnapshot already calls WaitForGVRDeletion()): 

err = DeleteAndWaitSnapshot(...)
switch (policy) {
case delete: framework.ExpectNoError(err)
case retain: framework.ExpectError(err)
}